### PR TITLE
Feature: collect and report logs

### PR DESCRIPTION
**Pull request by Superskyyy**
This is a OSPP Summer 2021 project supervised by @Humbertzhang | apache/skywalking#7118

The feature implements optional log reporter functionalities in alignment with the SkyWalking Java agent. 

- Intercepts logs from Python logging module.
- Reports logs via a new temporary gRPC channel(to be removed in the future).
- Supports unformatted/ formatted logs with custom layout.
- Supports custom logging level threshold.
- Bumps up submodule to support log collection protocols.
- Bumps up skywalking-eyes and adds a config entry to ignore `.gitignore` during license checks.

### CODE REVIEW

File: skywalking/agent/protocol/grpc_log.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                    timeout = config.QUEUE_TIMEOUT - int(time() - start)  # type: int
+                    if timeout <= 0:  # this is to make sure we exit eventually instead of being fed continuously
+                        return
+                    log_data = queue.get(block=block, timeout=timeout)  # type: LogData
+                except Empty:
+                    return
+
+                queue.task_done()
+
+                logger.debug('Reporting Log %s', log_data)
```
Comment (Humbertzhang) [inline skywalking/agent/protocol/grpc_log.py:95]: No need to print the `log_data` object out.

File: skywalking/log/sw_logging.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        agent.archive_log(log_data)
+
+    Logger.handle = _sw_handle
+
+    def transform(record) -> str:
+        if config.log_grpc_reporter_formatted:
+            layout = config.log_grpc_reporter_layout  # type: str
+            if layout:
+                from logging import Formatter
+                formatter = Formatter(fmt=layout)
```
Comment (Humbertzhang) [inline skywalking/log/sw_logging.py:90]: Is there a way not to init a new formatter object at every time?

File: skywalking/config.py [hunk 3]
```diff
@@ -67,6 +67,19 @@
                          os.getenv('SW_AGENT_PROFILE_ACTIVE') == 'True' else False  # type: bool
 profile_task_query_interval = int(os.getenv('SW_PROFILE_TASK_QUERY_INTERVAL') or '20')
 
+# NOTE - Log reporting requires a separate channel, will merge in the future.
```
Comment (Humbertzhang) [inline skywalking/config.py:70]: What is this comment means?

**Comment by Humbertzhang** [comment]
Test it manually and it generally looks good to me. You should add test for it after this PR merged. 
@kezhenxu94 Should this feature was tested by using [skywalking-infra-e2e](https://github.com/apache/skywalking-infra-e2e) ?

**Comment by kezhenxu94** [comment]
> @kezhenxu94, I am wondering if this feature should be tested by using [skywalking-infra-e2e](https://github.com/apache/skywalking-infra-e2e) ? 

I think it's not mandatory, though it would be a bonus if @Superskyyy can adopt skywalking-infra-e2e in the new test.

**Comment by wu-sheng** [comment]
> Test it manually and it generally looks good to me. You should add test for it after this PR merged.
> @kezhenxu94, I am wondering if this feature should be tested by using [skywalking-infra-e2e](https://github.com/apache/skywalking-infra-e2e) ?

This is recommended, that mechanism makes sure we wouldn't break it in the future.

File: docs/EnvVars.md [hunk 4]
```diff
@@ -32,3 +32,10 @@ Environment Variable | Description | Default
 | `SW_CELERY_PARAMETERS_LENGTH`| The maximum length of `celery` functions parameters, longer than this will be truncated, 0 turns off  | `512` |
 | `SW_AGENT_PROFILE_ACTIVE` | If `True`, Python agent will enable profile when user create a new profile task. Otherwise disable profile. | `False` |
 | `SW_PROFILE_TASK_QUERY_INTERVAL` | The number of seconds between two profile task query. | `20` |
+| `SW_AGENT_LOG_REPORTER_ACTIVE` | If `True`, Python agent will report collected logs to the OAP or Satellite. Otherwise, it disables the feature. | `False` |
```
Comment (wu-sheng) [inline docs/EnvVars.md:35]: Besides this doc change, let's add a specific doc to show users how to use it.

**Comment by wu-sheng** [comment]
And after this change, a new section should be added into main repo's doc, https://github.com/apache/skywalking/blob/master/docs/en/setup/backend/log-analyzer.md#java-agents-toolkits, which should link to this required doc https://github.com/apache/skywalking-python/pull/147#discussion_r684698040

File: skywalking/agent/protocol/grpc_log.py (hunk 1 above)
Comment (Superskyyy) [reply skywalking/agent/protocol/grpc_log.py:95]: Ok I will remove the argument.

File: skywalking/log/sw_logging.py (hunk 2 above)
Comment (Superskyyy) [reply skywalking/log/sw_logging.py:90]: There certainly is! Thanks for pointing that out. I'm moving the initialization part to the outer scope.

**Comment by Superskyyy** [comment]
> > Test it manually and it generally looks good to me. You should add test for it after this PR merged.
> > @kezhenxu94, I am wondering if this feature should be tested by using [skywalking-infra-e2e](https://github.com/apache/skywalking-infra-e2e) ?
> 
> This is recommended, that mechanism makes sure we wouldn't break it in the future.

Will do that next :)

File: docs/EnvVars.md (hunk 4 above)
Comment (Superskyyy) [reply docs/EnvVars.md:35]: @wu-sheng I just added a detailed guide, please see if that is sufficient.

**Review by Humbertzhang — APPROVED** [review]
LGTM, Thank you for your contributing!

**Review by kezhenxu94 — APPROVED** [review]
@Superskyyy nice work!! Thanks 🙇🏻

**Review by wu-sheng — APPROVED** [review]
Thanks. Please update the main repo doc linking to this doc.

**Comment by Superskyyy** [comment]
> Thanks. Please update the main repo doc linking to this doc.

Certainly.

**Comment by Humbertzhang** [comment]
Closes https://github.com/apache/skywalking/issues/7118

**Comment by tom-pytel** [comment]
So this is grpc only? No http or kafka protocol?

**Comment by kezhenxu94** [comment]
> So this is grpc only? No http or kafka protocol?

For this PR, yes, we need iterations to support http and kafka protocol

**Comment by tom-pytel** [comment]
> For this PR, yes, we need iterations to support http and kafka protocol

This functionality should be pushed into the protocols themselves, which would also remove the separate channel for grpc as the author said.

**Comment by kezhenxu94** [comment]
> This functionality should be pushed into the protocols themselves, which would also remove the separate channel for grpc as the author said.

@Humbertzhang will be working to remove the separate channel for logs by reusing the same gRPC protocol / channel.

**Pull request merged**